### DISCUSSION TITLE
Building a seedtool that runs in a web browser

### DISCUSSION THREAD

**Discussion by Sjlver** [BlockchainCommons/Gordian-Developer-Community#81] [General & Announcements]
Hello everybody out there using seedtool!

Since I've discovered Blockchain Commons and its projects like ByteWords, seedtool, UR, and SSKR, I've been fascinated by the clarity and simplicity of the ideas. These are tools that I've long wanted to have. Seedtool in particular would be fantastic to safely generate secret key shares, if only I could be confident about running it securely.

What I want is a seedtool that anyone can easily run on an airgapped device. That is why I am proposing to **build a seedtool that runs in a standalone HTML/JavaScript page** in a web browser. This means that any cheap old smartphone or laptop can be used to run it, without any prerequisites and without having to compile anything. I am looking for motivated contributors who want to help me in this endeavor.

### The principles of this project

* Build a single-page app that can be run in any browser.
* It works completely offline. It can be run in an incognito window, with all traces deleted once the window is closed.
* Core features:
    * Generate seeds
    * Generate seed shares and show them as ByteWords, URs, or QR codes
    * Show seed as BIP-39 so it can be imported in a wallet
    * Reconstruct a seed from shares (This is a bit secondary, but I would like users to only need one single tool for handling their seeds)
* Security is paramount:
    * Code should be very readable, easy to inspect, with few dependencies
    * The page can warn the user if the device is not in airplane mode (by trying a network request) or not in incognito mode (by looking for old previously-stored cookies)
    * The page can include clear and simple instructions for users, e.g., what to do with their shares.

Besides the Blockchain Commons projects, another source of inspiration for this is https://github.com/iancoleman/bip39 -- That tool handles the mnemonic -> address conversion, whereas my proposed seedtool would handle seed -> mnemonic and seed <-> shares.

### Status and where to go from here

I've created a git repository at ➡️ https://github.com/Sjlver/seedpage where the tool will be developed.

If you'd like to contribute, simply open an issue or discussion topic on that repository. All sorts of contributions are welcome, even just ideas or suggestions.

**Comment by ChristopherA** [comment]
This is great 👍 

We have an upcoming internship program this summer and I was going to suggest something similar as a suggestion for them. The minor difference was to leverage web assembly like @wolfmcnally did for the static https://LifeHash.info webpage (or even start with a fork the repo at https://github.com/BlockchainCommons/lifehash.info )

**Reply by Sjlver** [reply]
I'm quite open to using WebAssembly. It might indeed be a faster way to get started.

Two thoughts:
1. There might be value in having two independent implementations, at least if we can check automatically that they produce the same output.
2. I think the WebAssembly version would be a little harder for people to inspect and analyze.  On the other hand, the WebAssembly blob would have a clearly defined interface with the rest of the page, and that at least could be inspected. I would also assume that the entropy comes from the JavaScript side (I don't think WebAssembly can access the browser's RNG directly), so that crucial part would be explicit.

Whatever the implementation, I think it would be nice to have an integration test suite or so, which checks that the page generates exactly the same output as seedtool-cli.

For me, this is just a hobby, won't be big and professional like gnu ;-) and so I might not be very far into the project when summer comes and your intern starts. I'd be more than happy to collaborate with an intern when the time comes (or with anyone else, for that matter).

**Reply by Sjlver** [reply]
This was easier than I thought :) **Check out https://blog.purpureus.net/seedpage/**

Source code at https://github.com/Sjlver/seedpage

The current version is very basic, but demonstrates that seedtool can run inside a browser through WebAssembly. I'm hoping to spend some time over the coming weeks to polish seedpage.

Again: any sort of contribution or comment is welcome.

**Reply by Sjlver** [reply]
Now with updated look: https://blog.purpureus.net/seedpage/

**Reply by wolfmcnally** [reply]
Awesome! 🎉

As I'm sure you're aware, I love the flexibility of being able to run tools like LifeHash entirely in the browser. You've made a great start on making seedtool available to many people who otherwise wouldn't get to use it. Thank you for your work.

**Reply by shannona** [reply]
Definitely agreed that it's awesome, @Sjlver. If you're willing, I'd love to offer this in a Blockchain Commons repo. If you'd be willing to move it over, let me know!

**Reply by ChristopherA** [reply]
We probably should create a new repo with our secure template, and give @Sjlver maintainer privileges on it, rather than move his over. We should probably also parallel the Lifehash repo approach of where he puts the live web assembly LifeHash.info in a subfolder (I think). That way we can offer signed releases of the web page as well (which is particularly important with this project.)

@Sjlver could you coordinate with @shannona? His email is Shannon.Appelcline@gmail.com.

**Reply by shannona** [reply]
If that sounds OK to you, @Sjlver, I'll get it going.

**Reply by Sjlver** [reply]
@shannona cool for me. I've emailed you to coordinate.

**Reply by Sjlver** [reply]
> We should probably also parallel the Lifehash repo approach of where he puts the live web assembly LifeHash.info in a subfolder (I think).
> That way we can offer signed releases of the web page as well (which is particularly important with this project.)

This is how seedpage currently works, in fact. I copy it to the `/docs` folder, which then gets published through GitHub pages at https://blog.purpureus.net/seedpage/.

**Reply by shannona** [reply]
@Sjlver Great! I've created a "seedtool.info" repo and sent you an invite for maintain perms.

So, please feel free to populate it. A few notes:

1. As @ChristopherA suggests, we'd probably do well to parallel the lifehash repo organization (https://github.com/BlockchainCommons/lifehash.info) for ease of use.
2. We grabbed the seedtool.info domain, so it'd be great if we could set up a live version of the demo as well. @wolfmcnally can offer advice on how he did it with lifetool.info if needed.
3. Definitely integrate your README.md material with your standard README.md text. I can fill in the rest of the template if you'd like when you're done with your first pass, just let me know.
4. We need a signed CLA. That's at https://github.com/BlockchainCommons/seedtool.info/blob/master/CLA.md and should be put in the Signed-CLA directory.

If you have any questions or if anything isn't right, just let me know.

**Reply by Sjlver** [reply]
Quick update: I've pushed a new version to https://blog.purpureus.net/seedpage/ -- now with QR codes.

**Comment by ChristopherA** [comment]
This is great 👍 

I read you blog post — I presume you are aware of KeyTool-cli? https://github.com/BlockchainCommons/keytool-cli

Combination of both in Web Assembly could be really useful. 

Also let us know if you’d be interested in contributing this work as an official Blockchain Commons project with you as lead. Advantages are reviews of PRs, long-term support and integration as we update libraries, and we hope to be investing soon in third-party security reviews.

**Reply by Sjlver** [reply]
Thanks :) I'm glad that you like the project.

Yes, I am aware of keytool-cli. It might indeed be nice to integrate. I would just like to carefully think about the user experience and the security.
- Simplicity is important to me, and I'm not yet 100% sure that having one integrated (but more complex) tool is better than two separate ones. To me, the essential use case of seedpage is securely generating Shamir shares of an identity. For all the other parts I'd probably use separate tools, particularly hardware wallets.
- There are already a few browser-based wallets or keytool-like systems, e.g., https://github.com/iancoleman/bip39. These already support the "keytool in a browser" use case to some extent.

> Also let us know if you’d be interested in contributing this work as an official Blockchain Commons project with you as lead. 

Absolutely. The caveat being that this is a hobby project for me. I'm doing it completely in my free time and won't be able to invest more than a handful of hours per week. As it says in the license, `this software is provided by the copyright holders and contributors "as is" and any express or implied warranties, including, ...` the implied warranty of it being maintained for a long period of time, is disclaimed ;-) But if this condition is OK for you, then I'd be happy to make this an official part of Blockchain Commons.

**Reply by wolfmcnally** [reply]
Simplicity is super important, as is security. In my opinion, on the web accessing the general command line tool isn't as important as an interface targeted toward specific user stories, the example you gave being using SSKR to shard or recover a seed used as an identity.
